### PR TITLE
new Trigger and TriggerManager in executor

### PR DESCRIPTION
**Pull request by chengren311**
1. keep Trigger and TriggerManager in common module since they are still being used by flow scheduler.
2. Create a new KillExecutionAction using flowrunnermanager instead of executormanager to kill flow. Still keep the old one in common module for being compatible with existing SLA trigger in the database. Will remove the old one when all existing triggers expire.
3. Remove the dependency on ExecutorManager in SlaChecker, SlaAlertAction, replace it with executorloader. This will pave the way for moving these two classes into Executor module. Still keep them in common module for being compatible with existing SLA trigger in the database. Will move the old one when all existing triggers expire.

### CODE REVIEW

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+
+package azkaban.execapp;
+
+import azkaban.trigger.Condition;
+import azkaban.trigger.TriggerAction;
+import java.util.List;
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: java util import should be in the first section?

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  public long getNextCheckTime() {
+    return nextCheckTime;
+  }
+
+
+  /**
+   * Perform the action if trigger condition is met
+   *
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: extra line

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+   */
+  @Override
+  public void run() {
+    if(isTriggerExpired()) {
+      return ;
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: log?

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  @Inject
+  public TriggerManager() {
+    this.scheduledService = Executors.newScheduledThreadPool(SCHEDULED_THREAD_POOL_SIZE);
+  }
+
+  private Condition createCondition(SlaOption sla, int execId, String checkerName, String checkerMethod) {
+    SlaChecker slaFailChecker = new SlaChecker(checkerName, sla, execId);
+    Map<String, ConditionChecker> slaCheckers = new HashMap<>();
+    slaCheckers.put(slaFailChecker.getId(), slaFailChecker);
+    return new Condition(slaCheckers, slaFailChecker.getId() + "."+checkerMethod);
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: space between "+"

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  private List<TriggerAction> createActions(SlaOption sla, int execId) {
+    List<TriggerAction> actions = new ArrayList<>();
+    List<String> slaActions = sla.getActions();
+    for (String act : slaActions) {
+      TriggerAction action = null;
+      switch (act) {
+        case SlaOption.ACTION_ALERT:
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: swtich is a bit hacky here, should we use an TriggerActionFactory() for producing TriggerActions?

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      List<TriggerAction> actions = createActions(sla, execId);
+      Trigger trigger = new Trigger(execId, triggerCond, expireCond, actions);
+      long delay = trigger.getNextCheckTime() - System.currentTimeMillis();
+      logger.info("Adding sla trigger " + sla.toString() + " to execution " + execId + ", schedule to check in " + delay/1000 + " seconds");
+      scheduledService.schedule(trigger, delay, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  public void shutdown() {
+    scheduledService.shutdownNow();
```
Comment (shuzhang1989) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: do you want to have a shutdown hook on logging / executing scheduled but not executed tasks?

File: azkaban-common/src/main/java/azkaban/Constants.java [hunk 7]
```diff
@@ -37,6 +37,8 @@
 
   public static final String AZKABAN_SERVLET_CONTEXT_KEY = "azkaban_app";
 
+  // Internal username used as performing SLA action
```
Comment (gaofan0905) [inline azkaban-common/src/main/java/azkaban/Constants.java]: please update the comment "Internal username used to perform SLA action"

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.log4j.Logger;
+
+import azkaban.trigger.TriggerAction;
+
+public class KillExecutionAction implements TriggerAction {
+
+  public static final String type = "KillExecutionAction";
```
Comment (gaofan0905) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: Normally static final param name should all be capitalized according to LinkedIn coding convention. 

Also please check if this one need to be public or not.

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+  @SuppressWarnings("unchecked")
+  public static KillExecutionAction createFromJson(Object obj) {
+    return createFromJson((HashMap<String, Object>) obj);
+  }
+
+  public static KillExecutionAction createFromJson(HashMap<String, Object> obj) {
+    Map<String, Object> jsonObj = (HashMap<String, Object>) obj;
+    String objType = (String) jsonObj.get("type");
+    if (!objType.equals(type)) {
```
Comment (gaofan0905) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: objType might be null? Better to use type.equals(objType) instead to avoid NPE

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    ServiceProvider.SERVICE_PROVIDER.getInstance(FlowRunnerManager.class).cancelFlow(execId, SlaChecker.USER);
+  }
+
+  @Override
+  public void setContext(Map<String, Object> context) {
+  }
+
+  @Override
+  public String getDescription() {
+    return type + " for " + execId;
```
Comment (gaofan0905) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: The description can be improved to make it more helpful

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2012 LinkedIn Corp.
```
Comment (kunkun-tang) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: should be 2017.

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.execapp.action;
+
+import azkaban.Constants;
+import azkaban.ServiceProvider;
+import azkaban.execapp.FlowRunnerManager;
+import azkaban.trigger.builtin.SlaChecker;
```
Comment (kunkun-tang) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: never used.

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
+package azkaban.execapp;
```
Comment (kunkun-tang) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: License missing.

File: azkaban-common/src/main/java/azkaban/Constants.java (hunk 7 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/Constants.java]: +1

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 14]
```diff
@@ -15,7 +15,9 @@
  */
 package azkaban.executor;
 
+import azkaban.sla.SlaOption;
 import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: probably this import is not used. You can install the new style and run it on your changed files only. It may fix some of these cosmetic issues.

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java [hunk 15]
```diff
@@ -191,6 +202,10 @@ public void setSubmitTime(long submitTime) {
     this.submitTime = submitTime;
   }
 
+  public void setSlaOptions(List<SlaOption> slaOptions) {
+    this.slaOptions = slaOptions;
+  }
+
   @Override
   public Map<String, Object> toObject() {
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: This entire construct should probably be refactored as well. There are standard libraries like `gson` and `jackson` to do json serde. This is simply reinventing the wheel. 

If we do that then the `toObject` and `fillExecutableFromMapObject` methods interfaces can be removed and along with it, all the constants etc can be cleaned up as well.

File: azkaban-common/src/main/java/azkaban/trigger/builtin/SlaAlertAction.java [hunk 16]
```diff
@@ -39,17 +40,15 @@
   private SlaOption slaOption;
   private int execId;
   private AlerterHolder alerters;
-  private static ExecutorManagerAdapter executorManager;
+  private ExecutorLoader executorLoader;
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/trigger/builtin/SlaAlertAction.java]: Do these fields change? Else just make everyone final. Makes things simpler.

File: azkaban-common/src/main/java/azkaban/trigger/builtin/SlaChecker.java [hunk 17]
```diff
@@ -41,17 +42,14 @@
   private SlaOption slaOption;
   private int execId;
   private long checkTime = -1;
+  private ExecutorLoader executorLoader;
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/trigger/builtin/SlaChecker.java]: same comment as above. use final if you can.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 18]
```diff
@@ -629,16 +635,19 @@ public ExecutableFlow getExecutableFlow(int execId) {
...
+    FlowRunner flowRunner = (FlowRunner) event.getRunner();
+    ExecutableFlow flow = flowRunner.getExecutableFlow();
 
+    if (event.getType() == Event.Type.FLOW_FINISHED) {
       recentlyFinishedFlows.put(flow.getExecutionId(), flow);
       logger.info("Flow " + flow.getExecutionId()
           + " is finished. Adding it to recently finished flows list.");
       runningFlows.remove(flow.getExecutionId());
     }
+    else if (event.getType() == Event.Type.FLOW_STARTED) {
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: I suppose this is a logic change? What does this do?

Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: when a flow starts, add it's associated sla trigger to the trigger manager

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+    boolean isTriggerConditionMet = triggerCondition.isMet();
+
+    if (isTriggerConditionMet) {
+      logger.info("Condition " + triggerCondition.getExpression() + " met");
+      for (TriggerAction action : actions) {
+        try {
+          action.doAction();
+        } catch (Exception e) {
+          logger.error("Failed to do action " + action.getDescription() + " for execution " + azkaban.execapp.Trigger.this.execId, e);
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: you can probably reformat this.

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+      this.isExpired = true;
+    }
+  }
+
+  /**
+   * Check if the trigger is expired and reset isExpired
+   * @return true if trigger is expired
+   */
+  public boolean isTriggerExpired() {
+    isExpired = isExpired || expireCondition.isMet();
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: This is counterintuitive. Why do we need to reassign `isExpired`? Can `expireCondition.isMet()` become `false` after returning `true`?

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+      TriggerAction action = null;
+      switch (act) {
+        case SlaOption.ACTION_ALERT:
+          action = new SlaAlertAction("slaAlert", sla, execId);
+          break;
+        case SlaOption.ACTION_CANCEL_FLOW:
+          action = new KillExecutionAction("killExecution", execId);
+          break;
+        default:
+          break;
```
Comment (suvodeep-pyne) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: can we log this if we see any funny strings? 
It is probably cleaner to have enums here.

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java (hunk 8 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: i'm following legacy convention here for being consistent. I think action implementation needs to be refactor and type should be enum. but it will requires changes on multiple classes and would be too big to fit into this PR.

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java (hunk 11 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java (hunk 12 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: updated

**Comment by suvodeep-pyne** [comment]
Had an initial look at it. Seems good. Have some comments.

> Create a new KillExecutionAction using flowrunnermanager instead of executormanager to kill flow. Still keep the old one in common module for being compatible with existing SLA trigger in the database. Will remove the old one when all existing triggers expire.

Please mark the older classes `@deprecated` in that case. It will be good to add these comments in those classes to in javadoc, etc.

Some questions on testing:
 - What is your testing strategy for this change?
 - Can there be some automated tests added to help ensuring the reliability of the changes?

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java (hunk 13 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: updated

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java (hunk 15 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: agree, submited an issue request for tracking https://github.com/azkaban/azkaban/issues/1098

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 19 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: udpated

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java (hunk 21 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: right, open up a issue request https://github.com/azkaban/azkaban/issues/1099 for tracking, this should be another PR.

**Comment by chengren311** [comment]
@suvodeep-pyne 
add the deprecated annotation. 
Regarding my testing strategy, mostly I tested it manually by setting up different types of SLA on schedules and checkin if it works as expected.

Currently SLA feature is not covered by any of our automation tests. We should definitely add it in our automation test framework. I will work with @jamiesjc on that.

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 20 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: isExpired is only served as cached value of expireCondition.isMet() here.
expireCondition.isMet() requires talking to db several times to check the flow status.

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java (hunk 6 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: not necessary. All SLA checker tasks should be stopped when shutting down executor.

File: azkaban-common/src/main/java/azkaban/Constants.java (hunk 7 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/Constants.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java (hunk 10 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: can you suggest how?

File: azkaban-common/src/main/java/azkaban/trigger/builtin/SlaAlertAction.java (hunk 16 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/trigger/builtin/SlaAlertAction.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 1 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 2 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java (hunk 3 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/Trigger.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java (hunk 5 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/TriggerManager.java]: consider this as a refactor of https://github.com/azkaban/azkaban/issues/1099

File: azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java (hunk 9 above)
Comment (chengren311) [inline azkaban-exec-server/src/main/java/azkaban/execapp/action/KillExecutionAction.java]: consider it as part of refactoring https://github.com/azkaban/azkaban/issues/1098

File: azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java (hunk 14 above)
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java]: updated
